File: test_zerocode_lib.py
"""Library functions — no trickle imports.
Instrumented externally via: python -m trickle.auto_run
"""


def parse_csv(text, delimiter=","):
    lines = text.strip().split("\n")
    headers = [h.strip() for h in lines[0].split(delimiter)]
    rows = []
    for line in lines[1:]:
        vals = [v.strip() for v in line.split(delimiter)]
        rows.append(dict(zip(headers, vals)))
    return {"headers": headers, "rows": rows, "row_count": len(rows)}


def slugify(text, separator="-"):
    import re
    slug = text.lower()
    slug = re.sub(r"[^a-z0-9]+", separator, slug)
    slug = slug.strip(separator)
    return {"original": text, "slug": slug, "length": len(text)}


def merge_config(defaults, overrides):
    merged = {**defaults}
    for key, val in overrides.items():
        if val is not None:
            merged[key] = val
    return {
        "config": merged,
        "overridden_keys": [k for k in overrides if overrides[k] is not None],
        "total_keys": len(merged),
    }

File: test-zerocode-lib.js
/**
 * Library functions used by test-zerocode-app.js.
 * NOTE: This file has ZERO trickle imports — instrumentation happens
 * externally via `node -r trickle/auto`.
 */

function parseCSV(text, delimiter) {
  const lines = text.trim().split('\n');
  const headers = lines[0].split(delimiter || ',').map(h => h.trim());
  const rows = lines.slice(1).map(line => {
    const vals = line.split(delimiter || ',').map(v => v.trim());
    const obj = {};
    headers.forEach((h, i) => { obj[h] = vals[i] || ''; });
    return obj;
  });
  return { headers, rows, rowCount: rows.length };
}

function slugify(text, separator) {
  const sep = separator || '-';
  return {
    original: text,
    slug: text.toLowerCase().replace(/[^a-z0-9]+/g, sep).replace(new RegExp(`^${sep}|${sep}$`, 'g'), ''),
    length: text.length,
  };
}

function mergeConfig(defaults, overrides) {
  const merged = { ...defaults };
  for (const [key, val] of Object.entries(overrides)) {
    if (val !== undefined && val !== null) {
      merged[key] = val;
    }
  }
  return {
    config: merged,
    overriddenKeys: Object.keys(overrides).filter(k => overrides[k] !== undefined),
    totalKeys: Object.keys(merged).length,
  };
}

module.exports = { parseCSV, slugify, mergeConfig };
